### PR TITLE
new-log-viewer: Implement MenuBar and StatusBar with JoyUI components; add JoyUI custom app theme.

### PR DESCRIPTION
**Pull request by Henry8192**
# References
new-log-viewer series: #45 #46 #48 #51 #52 #53 #54 #55 #56 #57

# Description
### App Theme
Use JoyUI components for Button, Select, and Input. In `App.js`, `CssVarsProvider` is added as a root provider to provide styles for components described above.

### Menu Bar
- Four navigation buttons: used for jumping to first, previous, next and last page.
- File open button: used for open a new file. 
- Settings button: used to set various options, like page size.
  - Theme toggle button: Toggle between light, dark and system.
- Tips button: Show keyboard shortcuts. Currently is a no-op.

### Status Bar
Display status message and a button displaying current log event number. `onClick`, the button will copy the permalink of the current line to clipboard.

# Validation performed
Open log viewer.

### Settings Test
Click settings (gear) button, toggle theme from light, system and dark, observe the color change except the monaco editor. Set page size and verify using navigation buttons.

### Navigation Test
Click the four navigation buttons. Observe the change showed in the monaco editor as well as logEventNum in browser's navigation bar. Make sure to set proper pageSize in settings.

### Open File Test
Click file open button, open a jsonl file locally, observe that monaco editor opens the provided log file.

### Copy Permalink To Clipboard Test
Click the blue button (Log Event [current log event num] of [total log events]), it will copy the current log event's permalink to the clipboard.

### CODE REVIEW

File: new-log-viewer/src/App.tsx
Comment (junhaoliao) [inline new-log-viewer/src/App.tsx:54]: Note they Microsoft fonts' licenses might not be compatible with yscope-log-viewer's license and we might need licenses to be able to integrate the Microsoft fonts when we want to derive a commercial product from yscope-log-viewer. 

For now, I think it's safe to keep `-apple-system, BlinkMacSystemFont, system-ui, Ubuntu, "Droid Sans", Roboto`. We will introduce a global style variable `--ylv-ui-font-family` in #56 which you might want to leverage after that PR is merged.

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:272]: > ESLint: Caution: `React` also has a named export `useState`. Check if you meant to write `import {useState} from 'react'` instead.(import/no-named-as-default-member)

```suggestion
    const [open, setOpen] = useState<boolean>(false);
```

Let's try to fix any lint errors before pushing to PRs now we have the ESLint automation pipeline enabled.

File: new-log-viewer/src/App.tsx
Comment (junhaoliao) [inline new-log-viewer/src/App.tsx:15]: Can we move this into another source file?

File: new-log-viewer/src/App.tsx
Comment (junhaoliao) [inline new-log-viewer/src/App.tsx:61]: Is there any reason to have the customizations be `md` size specific?

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:36]: Shall we import from the separate files (as done below)? Tree-shaking might not always work, and we might end up with a big bundle that has unneeded imports.

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:169]: It seems the button is in blue colour, which looks the same as the other buttons. Why are we setting the color as "success" here? If we want the success button to look different than the other ones, shall we be more specific in the theme customization?

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:178]: Shall we name this "Reset Default" now this is no longer debug input?

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:1]: Let's move the components into separate source files. Ideally, in a single source file there should be only one component.

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:223]:
```suggestion
```

File: new-log-viewer/src/utils/config.ts
Comment (junhaoliao) [inline new-log-viewer/src/utils/config.ts:91]: Let's add some comment to explain why we're not handling `CONFIG_KEY.THEME` here.

File: new-log-viewer/src/utils/config.ts
Comment (junhaoliao) [inline new-log-viewer/src/utils/config.ts:127]: ditto. Let's leave a comment for THEME here.

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [reply new-log-viewer/src/components/Layout.tsx:1]: Also, when possible, let's move the `style` props on the components into stylesheets.

**Review by junhaoliao — COMMENTED** [review]
Let's work on code organization before we proceed to the next iteration of review.

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:301]: Let's pull out the handlers into named functions.

File: new-log-viewer/src/App.tsx
Comment (Henry8192) [reply new-log-viewer/src/App.tsx:15]: Moved to `utils/theme.tsx`.

File: new-log-viewer/src/App.tsx
Comment (Henry8192) [reply new-log-viewer/src/App.tsx:61]: Nope. I think this is kept in our offline discussion for possible customization. Shall we remove it?

Comment (junhaoliao) [reply new-log-viewer/src/App.tsx:61]: Sure, let's remove it. If we do not need to check the size, we can simply provide the overrides in an object like 
```
root: {
    borderRadius: "2px",
    ...
},
```
right?

File: new-log-viewer/src/App.tsx
Comment (junhaoliao) [reply new-log-viewer/src/App.tsx:15]: Are the changes pushed?

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/ConfigForm.tsx:116]: Instead of using a table, would it be easier to write something like
```
{formFields.map((field, index) => (
    <FormControl>
      <FormLabel>{field.label}</FormLabel>
      <Input .../>
      <FormHelperText>{field.helperText}</FormHelperText>
    </FormControl>)
}
```

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/ConfigForm.tsx:121]:
```suggestion
                    Apply & Reload
```

File: new-log-viewer/src/components/Layout.css
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.css:1]: Any reason we want to add the file name in the stylesheet?
```suggestion
```

File: new-log-viewer/src/components/Layout.css
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.css:32]:
```suggestion
}

```

File: new-log-viewer/src/components/Layout.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.tsx:98]: Shall we move this into another source file?

File: new-log-viewer/src/utils/config.ts
Comment (junhaoliao) [inline new-log-viewer/src/utils/config.ts:72]:
```suggestion
        // Do NOT handle `CONFIG_KEY.THEME` here since is handled by `CssVarsProvider`.
```

File: new-log-viewer/src/utils/config.ts
Comment (junhaoliao) [inline new-log-viewer/src/utils/config.ts:107]:
```suggestion
        // Do NOT handle `CONFIG_KEY.THEME` here since is handled by `CssVarsProvider`.
```

File: new-log-viewer/src/utils/theme.tsx
Comment (junhaoliao) [inline new-log-viewer/src/utils/theme.tsx:45]: Could you try if this works as intended?
```suggestion
        body: "var(--ylv-ui-font-family)",
```

**Review by junhaoliao — COMMENTED** [review]
Let's address the posted comments. Most importantly, we want <Layout/> to
```
return (<>
<MenuBar/>
<Editor/>
<StatusBar/>
</>);
```

File: new-log-viewer/src/components/Layout.css
Comment (Henry8192) [reply new-log-viewer/src/components/Layout.css:1]: I thought it is a tradition to add these on top of css files. Maybe I used too much ChatGPT and their generated css code all starts with these file name headers XD.

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:118]: Yes, it makes more sense if we use the latter approach, as it is Layout's job to use the CssVarsProvider, not in our App.tsx.

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:6]: I'll just use path imports for now. Leave it to you to add the plugin.

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:121]: Primary. There's no "success" now.

File: new-log-viewer/src/components/Layout.css
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.css:20]:
```suggestion
    height: 32px;
```

We should probably declare some global variables in https://github.com/y-scope/yscope-log-viewer/blob/c939f7b55b55b42f65226470d5277b15ac484665/new-log-viewer/src/index.css#L11 and replace the references in `.menu-bar` and `.editor-container` to avoid inconsistencies.

File: new-log-viewer/src/utils/theme.tsx
Comment (junhaoliao) [inline new-log-viewer/src/utils/theme.tsx:1]: Would it make more sense to place this inside `<new-log-viewer-root>/src/components` since this is UI related?

File: new-log-viewer/src/components/Layout.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/Layout.tsx:103]: Sure. Will do

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/ConfigForm.tsx:19]: Do we plan to address this FIXME in the same PR?

Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:19]: I don't really know. Do we need a helper text here? Your previous suggestion has one helper text field, but currently we haven't implement that.

File: new-log-viewer/src/utils/config.ts
Comment (junhaoliao) [inline new-log-viewer/src/utils/config.ts:107]: (ditto) Can we leave the `case CONFIG_KEY.THEME:` condition? That way IDEs / linters would not complain about any missing handling for that specific case. e.g.,

```
case CONFIG_KEY.THEME:
    // Handled by JoyUI instead.
    break;
```

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:171]: Shall we make it fullscreen?
```suggestion
                <ModalDialog layout={"fullscreen"}>
```

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:19]: nvm. I will add helper text field in the formFields

Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:19]: let's discuss offline about what exactly the helper texts should be

File: new-log-viewer/src/components/ConfigForm.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/ConfigForm.tsx:123]: I mean, we don't want to use success because I once overloaded the "success" property and make it color blue. The button should be "primary", so that the color is blue.

File: new-log-viewer/src/components/MenuBar.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/MenuBar.tsx]: Well, it turns out fullscreen is overwhelming because it takes too much space for only a few info. I think reverting this back will be better.
image

Comment (junhaoliao) [reply new-log-viewer/src/components/MenuBar.tsx]: Sorry I might have missed this in the previous comments:
Why do we need `orientation={"horizontal"}` on the `FormControl` component?

File: new-log-viewer/src/App.tsx
Comment (junhaoliao) [inline new-log-viewer/src/App.tsx:19]:
```suggestion
        <UrlContextProvider>
            <StateContextProvider>
                <Layout/>
            </StateContextProvider>
        </UrlContextProvider>
```

File: new-log-viewer/src/components/Layout.css
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.css:3]:
```suggestion
.layout {
```

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:45]: Module exports should be move to the end of the file. e.g.,

```
export {
    handleAction,
...
};

```

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:73]: ditto

File: new-log-viewer/src/components/StatusBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/StatusBar.tsx:14]: Module exports should be moved to the end of the file.

```
export default StatusBar;
```

Comment (junhaoliao) [inline new-log-viewer/src/components/StatusBar.tsx:14]:
```suggestion
const StatusBar = ({handleCopyLinkButtonClick, logEventNum, numEvents}: StatusBarProps) => {
```

Then we don't need the 
```
import React from "react";
```

File: new-log-viewer/src/components/StatusBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/StatusBar.tsx:26]: Shall we copy the function definition to this source file to avoid passing the callback as a prop to the component?

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:45]: Shall we move the definition of this function to `src/utils/actions.ts`?

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:70]:
```suggestion

/**
```

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:73]: If we are not exporting other variables in this source file, we can write `export default MenuBar` at the end of this source file.

File: new-log-viewer/src/components/Layout.css
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.css:9]: Shall we move these to their dedicated stylesheets? e.g. `src/components/MenuBar/index.css`?

File: new-log-viewer/src/components/MenuBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar.tsx:1]: Can we create directory `src/components/MenuBar/`, move this file inside and rename it as `index.tsx`? Then we can create a dedicated stylesheet `index.css` inside the same directory, which can be imported by this source file.

File: new-log-viewer/src/components/StatusBar.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/StatusBar.tsx:1]: Can we create directory `src/components/StatusBar/`, move this file inside and rename it as `index.tsx`? Then we can create a dedicated stylesheet `index.css` inside the same directory, which can be imported by this source file.

File: new-log-viewer/src/components/Layout.css
Comment (junhaoliao) [inline new-log-viewer/src/components/Layout.css:23]: Shall we move this to the SettingsModal's dedicated stylesheet file?

**Review by junhaoliao — COMMENTED** [review]
Regarding the height overflow issue, I believe we can remove the `height: 100%;` specifications on `.drop-file-container` and `.drop-file-children`. After you clean those up, please remove unneeded classes / className props if necessary.

File: new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx:181]: We don't have to pass the input props via `slotProps`, right?
```suggestion
                            <Input
                                defaultValue={field.initialValue}
                                name={field.name}
                                type={field.type}/>
```

File: new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx
Comment (junhaoliao) [inline new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx:127]:
```suggestion
                minWidth={"md"}
```

File: new-log-viewer/src/components/MenuBar.tsx
Comment (Henry8192) [reply new-log-viewer/src/components/MenuBar.tsx:1]: Any reasons of why we use index.css instead of menubar.css?

File: new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx
Comment (junhaoliao) [reply new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx:127]: @Henry8192 Shall we apply this? Currently the modal window is too narrow that the Format String input box cannot be shown fully.
image

Comment (Henry8192) [reply new-log-viewer/src/components/modals/SettingsModal/ConfigDialog.tsx:127]: Yup let's do this.

**Review by junhaoliao — APPROVED** [review]
lgtm

**Pull request merged**
